File: Packages/NoopRemoteSync/Sources/NoopRemoteSync/ManagedDocumentEnvelope.swift
import CryptoKit
import Foundation

public enum ManagedDocumentEnvelopeError: Error, Equatable {
    case invalidMetadata
    case invalidKey
    case invalidEnvelope
    case payloadTooLarge
    case authenticationFailed
}

public struct ManagedDocumentEnvelopeMetadata: Equatable, Sendable {
    public let accountScopeHash: String
    public let documentKind: ManagedDocumentKind
    public let documentID: UUID
    public let revision: Int64

    public init(
        accountScopeHash: String,
        documentKind: ManagedDocumentKind,
        documentID: UUID,
        revision: Int64
    ) throws {
        guard accountScopeHash.range(
            of: #"^[0-9a-f]{64}$"#,
            options: .regularExpression
        ) != nil,
        revision > 0 else {
            throw ManagedDocumentEnvelopeError.invalidMetadata
        }
        self.accountScopeHash = accountScopeHash
        self.documentKind = documentKind
        self.documentID = documentID
        self.revision = revision
    }

    public var authenticatedData: Data {
        Data(
            (
                "noop-managed-document-aad-v1\0"
                    + accountScopeHash + "\0"
                    + documentKind.rawValue + "\0"
                    + documentID.uuidString.lowercased() + "\0"
                    + String(revision)
            ).utf8
        )
    }
}

public enum ManagedDocumentEnvelope {
    public static let maximumEnvelopeBytes = 1_048_576
    public static let maximumPlaintextBytes = maximumEnvelopeBytes - 40

    private static let magic = Data([
        0x4e, 0x4f, 0x4f, 0x50, 0x44, 0x4f, 0x43, 0x00,
    ])

    public static func seal(
        _ plaintext: Data,
        key: Data,
        metadata: ManagedDocumentEnvelopeMetadata
    ) throws -> Data {
        try seal(plaintext, key: key, metadata: metadata, nonce: nil)
    }

    static func seal(
        _ plaintext: Data,
        key: Data,
        metadata: ManagedDocumentEnvelopeMetadata,
        nonce: Data?
    ) throws -> Data {
        guard key.count == 32 else {
            throw ManagedDocumentEnvelopeError.invalidKey
        }
        guard plaintext.count <= maximumPlaintextBytes else {
            throw ManagedDocumentEnvelopeError.payloadTooLarge
        }
        let aesNonce: AES.GCM.Nonce
        do {
            aesNonce = try nonce.map(AES.GCM.Nonce.init(data:)) ?? AES.GCM.Nonce()
        } catch {
            throw ManagedDocumentEnvelopeError.invalidEnvelope
        }
        let box: AES.GCM.SealedBox
        do {
            box = try AES.GCM.seal(
                plaintext,
                using: SymmetricKey(data: key),
                nonce: aesNonce,
                authenticating: metadata.authenticatedData
            )
        } catch {
            throw ManagedDocumentEnvelopeError.invalidEnvelope
        }
        var envelope = magic
        envelope.append(1)
        envelope.append(12)
        envelope.append(contentsOf: [0, 0])
        envelope.append(contentsOf: aesNonce)
        envelope.append(box.ciphertext)
        envelope.append(box.tag)
        guard envelope.count <= maximumEnvelopeBytes else {
            throw ManagedDocumentEnvelopeError.payloadTooLarge
        }
        return envelope
    }

    public static func open(
        _ envelope: Data,
        key: Data,
        metadata: ManagedDocumentEnvelopeMetadata
    ) throws -> Data {
        guard key.count == 32 else {
            throw ManagedDocumentEnvelopeError.invalidKey
        }
        guard envelope.count >= 40,
              envelope.count <= maximumEnvelopeBytes,
              envelope.prefix(magic.count) == magic,
              envelope[8] == 1,
              envelope[9] == 12,
              envelope[10] == 0,
              envelope[11] == 0 else {
            throw ManagedDocumentEnvelopeError.invalidEnvelope
        }
        let nonce = envelope.subdata(in: 12..<24)
        let ciphertext = envelope.subdata(in: 24..<(envelope.count - 16))
        let tag = envelope.suffix(16)
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce),
                ciphertext: ciphertext,
                tag: tag
            )
            return try AES.GCM.open(
                box,
                using: SymmetricKey(data: key),
                authenticating: metadata.authenticatedData
            )
        } catch {
            throw ManagedDocumentEnvelopeError.authenticationFailed
        }
    }
}

public struct ManagedWrappedDocumentKey: Codable, Equatable, Sendable {
    public let version: Int
    public let keyID: UUID
    public let wrappingKeyID: UUID
    public let wrappingRevision: Int
    public let algorithm: String
    public let wrappedKeyBase64: String
    public let wrappedKeySHA256: String

    public init(
        keyID: UUID,
        wrappingKeyID: UUID,
        wrappingRevision: Int,
        wrappedKey: Data
    ) throws {
        guard wrappingRevision > 0,
              wrappedKey.count == 72 else {
            throw ManagedDocumentEnvelopeError.invalidMetadata
        }
        self.version = 1
        self.keyID = keyID
        self.wrappingKeyID = wrappingKeyID
        self.wrappingRevision = wrappingRevision
        self.algorithm = "A256GCM"
        self.wrappedKeyBase64 = wrappedKey.base64EncodedString()
        self.wrappedKeySHA256 = ManagedDocumentEnvelopeDigest.sha256(wrappedKey)
    }

    public var wrappedKey: Data? {
        guard version == 1,
              algorithm == "A256GCM",
              wrappingRevision > 0,
              wrappedKeySHA256.range(
                  of: #"^[0-9a-f]{64}$"#,
                  options: .regularExpression
              ) != nil,
              let data = Data(base64Encoded: wrappedKeyBase64),
              data.count == 72,
              data.base64EncodedString() == wrappedKeyBase64,
              ManagedDocumentEnvelopeDigest.sha256(data) == wrappedKeySHA256
        else {
            return nil
        }
        return data
    }
}

public enum ManagedAccountMasterKeyBinding {
    private static let context =
        "noop-managed-account-master-key-confirmation-v1"
    private static let recoveryMethods = Set([
        "recovery_key",
        "device_transfer",
        "platform_escrow",
    ])

    public static func confirmationHMACSHA256(
        masterKey: Data,
        accountScopeHash: String,
        keyID: UUID,
        wrappingRevision: Int,
        wrappedKeySHA256: String,
        recoveryMethod: String
    ) throws -> String {
        guard masterKey.count == 32 else {
            throw ManagedDocumentEnvelopeError.invalidKey
        }
        guard accountScopeHash.range(
            of: #"^[0-9a-f]{64}$"#,
            options: .regularExpression
        ) != nil,
        wrappingRevision > 0,
        wrappedKeySHA256.range(
            of: #"^[0-9a-f]{64}$"#,
            options: .regularExpression
        ) != nil,
        recoveryMethods.contains(recoveryMethod) else {
            throw ManagedDocumentEnvelopeError.invalidMetadata
        }
        let message = [
            context,
            accountScopeHash,
            keyID.uuidString.lowercased(),
            String(wrappingRevision),
            wrappedKeySHA256,
            recoveryMethod,
        ].joined(separator: "\0")
        let confirmation = HMAC<SHA256>.authenticationCode(
            for: Data(message.utf8),
            using: SymmetricKey(data: masterKey)
        )
        return confirmation.map { String(format: "%02x", $0) }.joined()
    }

    public static func matches(
        masterKey: Data,
        accountScopeHash: String,
        receipt: ManagedAccountMasterKeyReceipt
    ) -> Bool {
        guard let expected = try? confirmationHMACSHA256(
            masterKey: masterKey,
            accountScopeHash: accountScopeHash,
            keyID: receipt.keyID,
            wrappingRevision: receipt.wrappingRevision,
            wrappedKeySHA256: receipt.wrappedKeySHA256,
            recoveryMethod: receipt.recoveryMethod
        ) else {
            return false
        }
        let lhs = Array(expected.utf8)
        let rhs = Array(receipt.masterKeyConfirmationHMACSHA256.utf8)
        guard lhs.count == rhs.count else { return false }
        var difference: UInt8 = 0
        for (left, right) in zip(lhs, rhs) {
            difference |= left ^ right
        }
        return difference == 0
    }
}

public struct ManagedAccountMasterKeyReceipt: Codable, Equatable, Sendable {
    public let version: Int
    public let keyID: UUID
    public let wrappingRevision: Int
    public let algorithm: String
    public let wrappedKeyBase64: String
    public let wrappedKeySHA256: String
    public let masterKeyConfirmationHMACSHA256: String
    public let recoveryMethod: String
    public let status: String

    public init(
        keyID: UUID,
        wrappingRevision: Int,
        wrappedKey: Data,
        masterKeyConfirmationHMACSHA256: String,
        recoveryMethod: String,
        status: String = "active"
    ) throws {
        guard wrappingRevision > 0,
              (40...16_384).contains(wrappedKey.count),
              masterKeyConfirmationHMACSHA256.range(
                  of: #"^[0-9a-f]{64}$"#,
                  options: .regularExpression
              ) != nil,
              ["recovery_key", "device_transfer", "platform_escrow"]
                .contains(recoveryMethod),
              status == "active" else {
            throw ManagedDocumentEnvelopeError.invalidMetadata
        }
        self.version = 1
        self.keyID = keyID
        self.wrappingRevision = wrappingRevision
        self.algorithm = "A256GCM"
        self.wrappedKeyBase64 = wrappedKey.base64EncodedString()
        self.wrappedKeySHA256 = ManagedDocumentEnvelopeDigest.sha256(
            wrappedKey
        )
        self.masterKeyConfirmationHMACSHA256 =
            masterKeyConfirmationHMACSHA256
        self.recoveryMethod = recoveryMethod
        self.status = status
    }

    public var wrappedKey: Data? {
        guard version == 1,
              algorithm == "A256GCM",
              wrappingRevision > 0,
              ["recovery_key", "device_transfer", "platform_escrow"]
                .contains(recoveryMethod),
              status == "active",
              wrappedKeySHA256.range(
                  of: #"^[0-9a-f]{64}$"#,
                  options: .regularExpression
              ) != nil,
              masterKeyConfirmationHMACSHA256.range(
                  of: #"^[0-9a-f]{64}$"#,
                  options: .regularExpression
              ) != nil,
              let data = Data(base64Encoded: wrappedKeyBase64),
              data.base64EncodedString() == wrappedKeyBase64,
              (40...16_384).contains(data.count),
              ManagedDocumentEnvelopeDigest.sha256(data)
                == wrappedKeySHA256 else {
            return nil
        }
        return data
    }
}

public struct ManagedDocumentKeyRotationPlan: Equatable, Sendable {
    public let planID: UUID
    public let accountMasterReceipt: ManagedAccountMasterKeyReceipt
    public let wrappedDocumentKeys: [ManagedWrappedDocumentKey]

    init(
        planID: UUID,
        accountMasterReceipt: ManagedAccountMasterKeyReceipt,
        wrappedDocumentKeys: [ManagedWrappedDocumentKey]
    ) {
        self.planID = planID
        self.accountMasterReceipt = accountMasterReceipt
        self.wrappedDocumentKeys = wrappedDocumentKeys
    }
}

enum ManagedDocumentKeyWrapEnvelope {
    private static let magic = Data([
        0x4e, 0x4f, 0x4f, 0x50, 0x4b, 0x45, 0x59, 0x00,
    ])

    static func seal(
        documentKey: Data,
        accountScopeHash: String,
        keyID: UUID,
        wrappingKeyID: UUID,
        wrappingRevision: Int,
        wrappingKey: Data,
        nonce: Data? = nil
    ) throws -> Data {
        guard documentKey.count == 32, wrappingKey.count == 32 else {
            throw ManagedDocumentEnvelopeError.invalidKey
        }
        let aad = authenticatedData(
            accountScopeHash: accountScopeHash,
            keyID: keyID,
            wrappingKeyID: wrappingKeyID,
            wrappingRevision: wrappingRevision
        )
        let aesNonce: AES.GCM.Nonce
        do {
            aesNonce = try nonce.map(AES.GCM.Nonce.init(data:)) ?? AES.GCM.Nonce()
        } catch {
            throw ManagedDocumentEnvelopeError.invalidEnvelope
        }
        let box = try AES.GCM.seal(
            documentKey,
            using: SymmetricKey(data: wrappingKey),
            nonce: aesNonce,
            authenticating: aad
        )
        var envelope = magic
        envelope.append(1)
        envelope.append(12)
        envelope.append(contentsOf: [0, 0])
        envelope.append(contentsOf: aesNonce)
        envelope.append(box.ciphertext)
        envelope.append(box.tag)
        return envelope
    }

    static func open(
        _ envelope: Data,
        accountScopeHash: String,
        keyID: UUID,
        wrappingKeyID: UUID,
        wrappingRevision: Int,
        wrappingKey: Data
    ) throws -> Data {
        guard wrappingKey.count == 32,
              envelope.count == 72,
              envelope.prefix(magic.count) == magic,
              envelope[8] == 1,
              envelope[9] == 12,
              envelope[10] == 0,
              envelope[11] == 0 else {
            throw ManagedDocumentEnvelopeError.invalidEnvelope
        }
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: envelope.subdata(in: 12..<24)),
                ciphertext: envelope.subdata(in: 24..<56),
                tag: envelope.suffix(16)
            )
            return try AES.GCM.open(
                box,
                using: SymmetricKey(data: wrappingKey),
                authenticating: authenticatedData(
                    accountScopeHash: accountScopeHash,
                    keyID: keyID,
                    wrappingKeyID: wrappingKeyID,
                    wrappingRevision: wrappingRevision
                )
            )
        } catch {
            throw ManagedDocumentEnvelopeError.authenticationFailed
        }
    }

    private static func authenticatedData(
        accountScopeHash: String,
        keyID: UUID,
        wrappingKeyID: UUID,
        wrappingRevision: Int
    ) -> Data {
        Data(
            (
                "noop-managed-document-key-wrap-v1\0"
                    + accountScopeHash + "\0"
                    + keyID.uuidString.lowercased() + "\0"
                    + wrappingKeyID.uuidString.lowercased() + "\0"
                    + String(wrappingRevision)
            ).utf8
        )
    }
}

enum ManagedDocumentEnvelopeDigest {
    static func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

File: android/app/src/main/java/com/noop/managed/ManagedDocumentEnvelope.kt
package com.noop.managed

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.UUID
import javax.crypto.AEADBadTagException
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

internal data class ManagedDocumentEnvelopeMetadata(
    val accountScopeHash: String,
    val documentKind: ManagedDocumentKind,
    val documentId: UUID,
    val revision: Long,
) {
    init {
        require(accountScopeHash.matches(SHA256))
        require(revision > 0L)
    }

    val authenticatedData: ByteArray
        get() = (
            "noop-managed-document-aad-v1\u0000$accountScopeHash\u0000" +
                "${documentKind.wireValue}\u0000${documentId.toString().lowercase()}\u0000$revision"
            ).toByteArray(StandardCharsets.UTF_8)

    private companion object {
        val SHA256 = Regex("^[0-9a-f]{64}$")
    }
}

internal object ManagedDocumentEnvelope {
    const val MAX_ENVELOPE_BYTES = 1_048_576
    const val MAX_PLAINTEXT_BYTES = MAX_ENVELOPE_BYTES - 40
    private val MAGIC = "NOOPDOC\u0000".toByteArray(StandardCharsets.US_ASCII)

    fun seal(
        plaintext: ByteArray,
        key: ByteArray,
        metadata: ManagedDocumentEnvelopeMetadata,
    ): ByteArray = seal(plaintext, key, metadata, null)

    internal fun seal(
        plaintext: ByteArray,
        key: ByteArray,
        metadata: ManagedDocumentEnvelopeMetadata,
        suppliedNonce: ByteArray?,
    ): ByteArray {
        require(key.size == 32)
        require(plaintext.size <= MAX_PLAINTEXT_BYTES)
        val nonce = suppliedNonce?.copyOf()
            ?: ByteArray(12).also(SecureRandom()::nextBytes)
        require(nonce.size == 12)
        val encrypted = Cipher.getInstance("AES/GCM/NoPadding").run {
            init(
                Cipher.ENCRYPT_MODE,
                SecretKeySpec(key, "AES"),
                GCMParameterSpec(128, nonce),
            )
            updateAAD(metadata.authenticatedData)
            doFinal(plaintext)
        }
        return MAGIC + byteArrayOf(1, 12, 0, 0) + nonce + encrypted
    }

    fun open(
        envelope: ByteArray,
        key: ByteArray,
        metadata: ManagedDocumentEnvelopeMetadata,
    ): ByteArray {
        require(key.size == 32)
        if (
            envelope.size !in 40..MAX_ENVELOPE_BYTES ||
            !envelope.copyOfRange(0, 8).contentEquals(MAGIC) ||
            envelope[8] != 1.toByte() ||
            envelope[9] != 12.toByte() ||
            envelope[10] != 0.toByte() ||
            envelope[11] != 0.toByte()
        ) {
            throw ManagedStorageException.InvalidResponse()
        }
        return try {
            Cipher.getInstance("AES/GCM/NoPadding").run {
                init(
                    Cipher.DECRYPT_MODE,
                    SecretKeySpec(key, "AES"),
                    GCMParameterSpec(128, envelope.copyOfRange(12, 24)),
                )
                updateAAD(metadata.authenticatedData)
                doFinal(envelope, 24, envelope.size - 24)
            }
        } catch (_: AEADBadTagException) {
            throw ManagedStorageException.InvalidResponse()
        } catch (_: java.security.GeneralSecurityException) {
            throw ManagedStorageException.InvalidResponse()
        }
    }
}

internal data class ManagedWrappedDocumentKey(
    val version: Int = 1,
    val keyId: UUID,
    val wrappingKeyId: UUID,
    val wrappingRevision: Int,
    val algorithm: String = "A256GCM",
    val wrappedKey: ByteArray,
) {
    init {
        require(version == 1)
        require(wrappingRevision > 0)
        require(algorithm == "A256GCM")
        require(wrappedKey.size == 72)
    }

    val wrappedKeySha256: String
        get() = sha256(wrappedKey)

    override fun equals(other: Any?): Boolean =
        other is ManagedWrappedDocumentKey &&
            version == other.version &&
            keyId == other.keyId &&
            wrappingKeyId == other.wrappingKeyId &&
            wrappingRevision == other.wrappingRevision &&
            algorithm == other.algorithm &&
            wrappedKey.contentEquals(other.wrappedKey)

    override fun hashCode(): Int =
        31 * keyId.hashCode() + wrappedKey.contentHashCode()
}

internal object ManagedAccountMasterKeyBinding {
    private const val CONTEXT =
        "noop-managed-account-master-key-confirmation-v1"
    private val SHA256 = Regex("^[0-9a-f]{64}$")
    private val RECOVERY_METHODS = setOf(
        "recovery_key",
        "device_transfer",
        "platform_escrow",
    )

    fun confirmationHmacSha256(
        masterKey: ByteArray,
        accountScopeHash: String,
        keyId: UUID,
        wrappingRevision: Int,
        wrappedKeySha256: String,
        recoveryMethod: String,
    ): String {
        require(masterKey.size == 32)
        require(accountScopeHash.matches(SHA256))
        require(wrappingRevision > 0)
        require(wrappedKeySha256.matches(SHA256))
        require(recoveryMethod in RECOVERY_METHODS)
        val message = listOf(
            CONTEXT,
            accountScopeHash,
            keyId.toString().lowercase(),
            wrappingRevision.toString(),
            wrappedKeySha256,
            recoveryMethod,
        ).joinToString("\u0000")
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(masterKey, "HmacSHA256"))
        return mac.doFinal(message.toByteArray(StandardCharsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
    }

    fun matches(
        masterKey: ByteArray,
        accountScopeHash: String,
        receipt: ManagedAccountMasterKeyReceipt,
    ): Boolean = runCatching {
        MessageDigest.isEqual(
            confirmationHmacSha256(
                masterKey = masterKey,
                accountScopeHash = accountScopeHash,
                keyId = receipt.keyId,
                wrappingRevision = receipt.wrappingRevision,
                wrappedKeySha256 = receipt.wrappedKeySha256,
                recoveryMethod = receipt.recoveryMethod,
            ).toByteArray(StandardCharsets.US_ASCII),
            receipt.masterKeyConfirmationHmacSha256
                .toByteArray(StandardCharsets.US_ASCII),
        )
    }.getOrDefault(false)
}

internal data class ManagedAccountMasterKeyReceipt(
    val version: Int = 1,
    val keyId: UUID,
    val wrappingRevision: Int,
    val algorithm: String = "A256GCM",
    val wrappedKey: ByteArray,
    val masterKeyConfirmationHmacSha256: String,
    val recoveryMethod: String,
    val status: String = "active",
) {
    init {
        require(version == 1)
        require(wrappingRevision > 0)
        require(algorithm == "A256GCM")
        require(wrappedKey.size in 40..16_384)
        require(masterKeyConfirmationHmacSha256.matches(SHA256))
        require(recoveryMethod in RECOVERY_METHODS)
        require(status == "active")
    }

    val wrappedKeySha256: String
        get() = sha256(wrappedKey)

    override fun equals(other: Any?): Boolean =
        other is ManagedAccountMasterKeyReceipt &&
            version == other.version &&
            keyId == other.keyId &&
            wrappingRevision == other.wrappingRevision &&
            algorithm == other.algorithm &&
            wrappedKey.contentEquals(other.wrappedKey) &&
            masterKeyConfirmationHmacSha256 ==
                other.masterKeyConfirmationHmacSha256 &&
            recoveryMethod == other.recoveryMethod &&
            status == other.status

    override fun hashCode(): Int =
        31 * keyId.hashCode() + wrappedKey.contentHashCode()

    private companion object {
        val SHA256 = Regex("^[0-9a-f]{64}$")
        val RECOVERY_METHODS = setOf(
            "recovery_key",
            "device_transfer",
            "platform_escrow",
        )
    }
}

internal data class ManagedDocumentKeyRotationPlan(
    val planId: UUID,
    val accountMasterReceipt: ManagedAccountMasterKeyReceipt,
    val wrappedDocumentKeys: List<ManagedWrappedDocumentKey>,
)

internal object ManagedDocumentKeyWrapEnvelope {
    private val MAGIC = "NOOPKEY\u0000".toByteArray(StandardCharsets.US_ASCII)

    fun seal(
        documentKey: ByteArray,
        accountScopeHash: String,
        keyId: UUID,
        wrappingKeyId: UUID,
        wrappingRevision: Int,
        wrappingKey: ByteArray,
        suppliedNonce: ByteArray? = null,
    ): ByteArray {
        require(documentKey.size == 32 && wrappingKey.size == 32)
        val nonce = suppliedNonce?.copyOf()
            ?: ByteArray(12).also(SecureRandom()::nextBytes)
        require(nonce.size == 12)
        val encrypted = Cipher.getInstance("AES/GCM/NoPadding").run {
            init(
                Cipher.ENCRYPT_MODE,
                SecretKeySpec(wrappingKey, "AES"),
                GCMParameterSpec(128, nonce),
            )
            updateAAD(aad(accountScopeHash, keyId, wrappingKeyId, wrappingRevision))
            doFinal(documentKey)
        }
        return MAGIC + byteArrayOf(1, 12, 0, 0) + nonce + encrypted
    }

    fun open(
        envelope: ByteArray,
        accountScopeHash: String,
        keyId: UUID,
        wrappingKeyId: UUID,
        wrappingRevision: Int,
        wrappingKey: ByteArray,
    ): ByteArray {
        if (
            wrappingKey.size != 32 ||
            envelope.size != 72 ||
            !envelope.copyOfRange(0, 8).contentEquals(MAGIC) ||
            !envelope.copyOfRange(8, 12).contentEquals(byteArrayOf(1, 12, 0, 0))
        ) {
            throw ManagedStorageException.InvalidResponse()
        }
        return try {
            Cipher.getInstance("AES/GCM/NoPadding").run {
                init(
                    Cipher.DECRYPT_MODE,
                    SecretKeySpec(wrappingKey, "AES"),
                    GCMParameterSpec(128, envelope.copyOfRange(12, 24)),
                )
                updateAAD(aad(accountScopeHash, keyId, wrappingKeyId, wrappingRevision))
                doFinal(envelope, 24, envelope.size - 24)
            }
        } catch (_: java.security.GeneralSecurityException) {
            throw ManagedStorageException.InvalidResponse()
        }
    }

    private fun aad(
        accountScopeHash: String,
        keyId: UUID,
        wrappingKeyId: UUID,
        wrappingRevision: Int,
    ): ByteArray = (
        "noop-managed-document-key-wrap-v1\u0000$accountScopeHash\u0000" +
            "${keyId.toString().lowercase()}\u0000" +
            "${wrappingKeyId.toString().lowercase()}\u0000$wrappingRevision"
        ).toByteArray(StandardCharsets.UTF_8)
}

internal fun sha256(value: ByteArray): String =
    MessageDigest.getInstance("SHA-256")
        .digest(value)
        .joinToString("") { "%02x".format(it) }
